File: android/app/src/main/kotlin/com/timecoin/wallet/network/PeerDiscovery.kt
package com.timecoin.wallet.network

import android.util.Log
import com.timecoin.wallet.model.PeerInfo
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.request.*
import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import java.io.File
import java.net.InetSocketAddress
import java.net.Socket

/**
 * Peer discovery — mirrors the desktop wallet's peer_discovery.rs + service.rs.
 *
 * Discovery flow:
 *   1. Collect candidates: manual peers (time.conf) + API peers + cached peers
 *   2. Parallel probe all candidates: TCP ping, RPC health check, WS availability
 *   3. Consensus filter: discard peers >3 blocks behind the best height
 *   4. Sort: healthy → WS-capable → lowest ping
 *   5. Gossip: ask healthy peers for their known peers, probe new discoveries
 *   6. Re-sort and return ranked list
 *   7. Cache healthy peers for offline fallback
 */
object PeerDiscovery {

    private const val TAG = "PeerDiscovery"
    private const val MAINNET_URL = "https://time-coin.io/api/peers"
    private const val TESTNET_URL = "https://time-coin.io/api/testnet/peers"
    private const val MAINNET_RPC_PORT = 24001
    private const val TESTNET_RPC_PORT = 24101
    private const val TCP_TIMEOUT_MS = 1000
    private const val HEALTH_TIMEOUT_MS = 4000L
    private const val WS_TIMEOUT_MS = 3000L
    private const val MAX_BLOCK_LAG = 3
    private const val MAX_PEERS = 8

    // ── Public API ──

    /**
     * Discover and rank peers. Returns a sorted list of [PeerInfo] —
     * best peer first (healthy, WS-capable, lowest latency).
     */
    suspend fun discoverAndRank(
        isTestnet: Boolean,
        manualEndpoints: List<String> = emptyList(),
        credentials: Pair<String, String>? = null,
        cacheDir: File? = null,
    ): List<PeerInfo> = coroutineScope {
        val rpcPort = if (isTestnet) TESTNET_RPC_PORT else MAINNET_RPC_PORT

        // 1. Collect candidate endpoints
        val candidates = mutableSetOf<String>()
        candidates.addAll(manualEndpoints)
        try {
            val apiPeers = fetchFromApi(isTestnet)
            candidates.addAll(apiPeers)
            if (cacheDir != null) saveCache(isTestnet, apiPeers, cacheDir)
        } catch (e: Exception) {
            Log.w(TAG, "API discovery failed: ${e.message}")
            if (cacheDir != null) {
                loadCache(isTestnet, cacheDir)?.let { candidates.addAll(it) }
            }
        }
        Log.d(TAG, "Collected ${candidates.size} candidate peers")

        if (candidates.isEmpty()) return@coroutineScope emptyList()

        // 2. Parallel probe all candidates
        var peerInfos = probeAll(candidates.toList(), credentials, rpcPort)
        Log.d(TAG, "Probed ${peerInfos.size} peers, ${peerInfos.count { it.isHealthy }} healthy")

        // 3. Consensus filter
        peerInfos = consensusFilter(peerInfos)

        // 4. Sort: healthy → WS → lowest ping
        peerInfos = rankPeers(peerInfos)

        // 5. Gossip discovery — ask healthy peers for their known peers
        val healthyPeers = peerInfos.filter { it.isHealthy }.take(3)
        val knownEndpoints = peerInfos.map { it.endpoint }.toSet()
        val gossipEndpoints = mutableSetOf<String>()

        for (peer in healthyPeers) {
            try {
                val client = MasternodeClient(peer.endpoint, credentials)
                val peerAddresses = withTimeout(HEALTH_TIMEOUT_MS) {
                    client.getPeerAddresses()
                }
                client.close()
                for (addr in peerAddresses) {
                    // addr is "IP:P2P_PORT" — extract IP, build RPC endpoint
                    val ip = addr.substringBefore(':')
                    val ep = "https://$ip:$rpcPort"
                    if (ep !in knownEndpoints) gossipEndpoints.add(ep)
                }
            } catch (e: Exception) {
                Log.w(TAG, "Gossip from ${peer.endpoint} failed: ${e.message}")
            }
        }

        if (gossipEndpoints.isNotEmpty()) {
            Log.d(TAG, "Gossip discovered ${gossipEndpoints.size} new peers")
            val gossipInfos = probeAll(gossipEndpoints.toList(), credentials, rpcPort)
            val filteredGossip = consensusFilter(gossipInfos, peerInfos)

            // Merge: replace slowest peers if gossip peers are faster
            val merged = (peerInfos + filteredGossip).distinctBy { it.endpoint }
            peerInfos = rankPeers(merged)
        }

        // 6. Cap at MAX_PEERS
        peerInfos = peerInfos.take(MAX_PEERS)

        // 7. Cache healthy peers
        if (cacheDir != null) {
            val healthyEndpoints = peerInfos.filter { it.isHealthy }.map { it.endpoint }
            if (healthyEndpoints.isNotEmpty()) {
                saveCache(isTestnet, healthyEndpoints, cacheDir)
            }
        }

        Log.d(TAG, "Final peer list (${peerInfos.size}): " +
            peerInfos.joinToString { "${it.endpoint} ping=${it.pingMs}ms h=${it.blockHeight}" })

        peerInfos
    }

    /**
     * Simple fetch for backward compatibility — returns endpoint strings only.
     */
    suspend fun fetchPeers(isTestnet: Boolean, cacheDir: File? = null): List<String> {
        return try {
            val peers = fetchFromApi(isTestnet)
            if (cacheDir != null) saveCache(isTestnet, peers, cacheDir)
            peers
        } catch (e: Exception) {
            if (cacheDir != null) loadCache(isTestnet, cacheDir) ?: emptyList()
            else emptyList()
        }
    }

    // ── Probing ──

    /**
     * Probe all endpoints in parallel. Each probe does:
     *   1. TCP connect to measure ping latency
     *   2. RPC health check (HTTPS → HTTP fallback)
     *   3. WebSocket availability check
     */
    private suspend fun probeAll(
        endpoints: List<String>,
        credentials: Pair<String, String>?,
        rpcPort: Int,
    ): List<PeerInfo> = coroutineScope {
        endpoints.map { endpoint ->
            async(Dispatchers.IO) { probePeer(endpoint, credentials, rpcPort) }
        }.awaitAll()
    }

    private suspend fun probePeer(
        endpoint: String,
        credentials: Pair<String, String>?,
        rpcPort: Int,
    ): PeerInfo {
        // Extract host/port for TCP ping
        val url = if (!endpoint.startsWith("http")) "https://$endpoint" else endpoint
        val host = url.removePrefix("https://").removePrefix("http://").substringBefore(':')
        val port = url.substringAfterLast(':').toIntOrNull() ?: rpcPort

        // 1. TCP ping
        val pingMs = measureTcpPing(host, port)

        // 2. RPC health check (try HTTPS first, then HTTP)
        var blockHeight: Long? = null
        var version: String? = null
        var isHealthy = false
        var isSyncing = false
        var workingUrl = url

        for (scheme in listOf("https", "http")) {
            val tryUrl = "$scheme://$host:$port"
            try {
                val client = MasternodeClient(tryUrl, credentials)
                val health = withTimeout(HEALTH_TIMEOUT_MS) { client.healthCheck() }
                client.close()
                blockHeight = health.blockHeight
                version = health.version
                isHealthy = true
                isSyncing = health.isSyncing
                workingUrl = tryUrl
                break
            } catch (e: Exception) {
                Log.v(TAG, "Health check $tryUrl failed: ${e.message}")
            }
        }

        // 3. WebSocket probe (RPC port + 1)
        val wsAvailable = if (isHealthy) {
            checkWsAvailable(host, port + 1)
        } else false

        return PeerInfo(
            endpoint = workingUrl,
            isActive = false,
            isHealthy = isHealthy,
            wsAvailable = wsAvailable,
            pingMs = pingMs,
            blockHeight = blockHeight,
            version = version,
            isSyncing = isSyncing,
        )
    }

    /** TCP connect to measure round-trip latency in milliseconds. */
    private fun measureTcpPing(host: String, port: Int): Long? {
        return try {
            val start = System.nanoTime()
            Socket().use { socket ->
                socket.connect(InetSocketAddress(host, port), TCP_TIMEOUT_MS)
            }
            (System.nanoTime() - start) / 1_000_000
        } catch (_: Exception) {
            null
        }
    }

    /** Check if WebSocket port is accepting connections. */
    private suspend fun checkWsAvailable(host: String, wsPort: Int): Boolean {
        return withContext(Dispatchers.IO) {
            try {
                withTimeout(WS_TIMEOUT_MS) {
                    Socket().use { socket ->
                        socket.connect(InetSocketAddress(host, wsPort), TCP_TIMEOUT_MS)
                    }
                    true
                }
            } catch (_: Exception) {
                false
            }
        }
    }

    // ── Filtering & Sorting ──

    /**
     * Consensus filter — discard peers whose block height is more than
     * [MAX_BLOCK_LAG] blocks behind the best known height.
     */
    private fun consensusFilter(
        peers: List<PeerInfo>,
        existingPeers: List<PeerInfo> = emptyList(),
    ): List<PeerInfo> {
        val allHeights = (peers + existingPeers).mapNotNull { it.blockHeight }
        if (allHeights.isEmpty()) return peers
        val bestHeight = allHeights.max()

        return peers.filter { p ->
            if (p.blockHeight == null) true // keep unhealthy peers (sorted last)
            else (bestHeight - p.blockHeight) <= MAX_BLOCK_LAG
        }.also { filtered ->
            val dropped = peers.size - filtered.size
            if (dropped > 0) Log.d(TAG, "Consensus filter dropped $dropped lagging peers (best=$bestHeight)")
        }
    }

    /** Sort peers: healthy first → not syncing → WS-capable → lowest ping. */
    private fun rankPeers(peers: List<PeerInfo>): List<PeerInfo> {
        return peers.sortedWith(
            compareByDescending<PeerInfo> { it.isHealthy }
                .thenBy { it.isSyncing }  // fully-synced peers before IBD peers
                .thenByDescending { it.wsAvailable }
                .thenBy { it.pingMs ?: Long.MAX_VALUE }
        )
    }

    // ── API & Cache ──

    private suspend fun fetchFromApi(isTestnet: Boolean): List<String> {
        val url = if (isTestnet) TESTNET_URL else MAINNET_URL
        val port = if (isTestnet) TESTNET_RPC_PORT else MAINNET_RPC_PORT

        val client = HttpClient(OkHttp)
        try {
            val response: String = client.get(url).body()
            val ips = Json.parseToJsonElement(response).jsonArray
                .mapNotNull { it.jsonPrimitive.contentOrNull }

            require(ips.isNotEmpty()) { "No peers returned from API" }
            return ips.map { ip -> "https://$ip:$port" }
        } finally {
            client.close()
        }
    }

    private fun saveCache(isTestnet: Boolean, peers: List<String>, cacheDir: File) {
        try {
            val file = File(cacheDir, cacheFilename(isTestnet))
            file.writeText(Json.encodeToString(JsonArray.serializer(), buildJsonArray {
                peers.forEach { add(it) }
            }))
        } catch (_: Exception) { }
    }

    private fun loadCache(isTestnet: Boolean, cacheDir: File): List<String>? {
        return try {
            val file = File(cacheDir, cacheFilename(isTestnet))
            if (!file.exists()) return null
            Json.parseToJsonElement(file.readText()).jsonArray
                .mapNotNull { it.jsonPrimitive.contentOrNull }
        } catch (_: Exception) { null }
    }

    private fun cacheFilename(isTestnet: Boolean) =
        if (isTestnet) "peers-testnet.json" else "peers-mainnet.json"
}

File: ios/TimeCoinWallet/Network/PeerDiscovery.swift
import Foundation
import Network

// Peer discovery — mirrors Android PeerDiscovery.kt.
// Flow: collect candidates → parallel probe → consensus filter → rank → gossip → cache
actor PeerDiscovery {
    private static let mainnetAPIURL = "https://time-coin.io/api/peers"
    private static let testnetAPIURL = "https://time-coin.io/api/testnet/peers"
    private static let mainnetRPCPort = 24001
    private static let testnetRPCPort = 24101
    private static let healthTimeoutSecs = 4.0
    private static let maxBlockLag: Int64 = 3
    private static let maxPeers = 8

    nonisolated static let shared = PeerDiscovery()

    private var blacklist: Set<String> = []

    func blacklistPeer(endpoint: String) {
        let host = endpoint
            .replacingOccurrences(of: "https://", with: "")
            .replacingOccurrences(of: "http://", with: "")
            .components(separatedBy: ":").first ?? endpoint
        blacklist.insert(host)
    }

    func discoverAndRank(
        isTestnet: Bool,
        manualEndpoints: [String] = [],
        credentials: (String, String)? = nil,
        filesDir: URL? = nil,
        expectedGenesisHash: String? = nil
    ) async -> [PeerInfo] {
        let rpcPort = isTestnet ? Self.testnetRPCPort : Self.mainnetRPCPort

        // 1. Collect candidates
        var candidates = Set<String>(manualEndpoints)
        do {
            let apiPeers = try await Self.fetchFromAPI(isTestnet: isTestnet, rpcPort: rpcPort)
            candidates.formUnion(apiPeers)
            if let dir = filesDir { Self.saveCache(isTestnet: isTestnet, peers: Array(apiPeers), dir: dir) }
        } catch {
            if let dir = filesDir, let cached = Self.loadCache(isTestnet: isTestnet, dir: dir) {
                candidates.formUnion(cached)
            }
        }

        let bl = blacklist
        candidates = candidates.filter { ep in
            let host = ep.replacingOccurrences(of: "https://", with: "")
                .replacingOccurrences(of: "http://", with: "")
                .components(separatedBy: ":").first ?? ep
            return !bl.contains(host)
        }

        if candidates.isEmpty { return [] }

        // 2. Parallel probe
        var peerInfos = await probeAll(
            endpoints: Array(candidates), credentials: credentials,
            rpcPort: rpcPort, expectedGenesisHash: expectedGenesisHash)

        // 3. Consensus filter
        peerInfos = Self.consensusFilter(peers: peerInfos)

        // 4. Sort
        peerInfos = Self.rankPeers(peerInfos)

        // 5. Gossip from top 3 healthy peers
        let healthyPeers = peerInfos.filter { $0.isHealthy }.prefix(3)
        let knownEndpoints = Set(peerInfos.map { $0.endpoint })
        var gossipEndpoints = Set<String>()

        for peer in healthyPeers {
            let client = MasternodeClient(endpoint: peer.endpoint, credentials: credentials)
            if let peerAddresses = try? await client.getPeerAddresses() {
                for addr in peerAddresses {
                    let ip = addr.components(separatedBy: ":").first ?? addr
                    if bl.contains(ip) { continue }
                    let ep = "https://\(ip):\(rpcPort)"
                    if !knownEndpoints.contains(ep) { gossipEndpoints.insert(ep) }
                }
            }
        }

        if !gossipEndpoints.isEmpty {
            let gossipInfos = await probeAll(
                endpoints: Array(gossipEndpoints), credentials: credentials,
                rpcPort: rpcPort, expectedGenesisHash: expectedGenesisHash)
            let filtered = Self.consensusFilter(peers: gossipInfos, existingPeers: peerInfos)
            let merged = Dictionary(grouping: peerInfos + filtered, by: { $0.endpoint })
                .compactMap { $0.value.first }
            peerInfos = Self.rankPeers(merged)
        }

        // 6. Cap at max
        peerInfos = Array(peerInfos.prefix(Self.maxPeers))

        // 7. Cache healthy peers
        if let dir = filesDir {
            let healthyEndpoints = peerInfos.filter { $0.isHealthy }.map { $0.endpoint }
            if !healthyEndpoints.isEmpty { Self.saveCache(isTestnet: isTestnet, peers: healthyEndpoints, dir: dir) }
        }

        return peerInfos
    }

    private func probeAll(
        endpoints: [String], credentials: (String, String)?,
        rpcPort: Int, expectedGenesisHash: String?
    ) async -> [PeerInfo] {
        await withTaskGroup(of: PeerInfo.self) { group in
            for ep in endpoints {
                group.addTask {
                    await self.probePeer(endpoint: ep, credentials: credentials,
                                        rpcPort: rpcPort, expectedGenesisHash: expectedGenesisHash)
                }
            }
            var results: [PeerInfo] = []
            for await info in group { results.append(info) }
            return results
        }
    }

    private func probePeer(
        endpoint: String, credentials: (String, String)?,
        rpcPort: Int, expectedGenesisHash: String?
    ) async -> PeerInfo {
        let url = endpoint.hasPrefix("http") ? endpoint : "https://\(endpoint)"
        let host = url.replacingOccurrences(of: "https://", with: "")
            .replacingOccurrences(of: "http://", with: "")
            .components(separatedBy: ":").first ?? url
        let port = Int(url.components(separatedBy: ":").last ?? "") ?? rpcPort

        let pingMs = await Self.measureTcpPing(host: host, port: port)

        var blockHeight: Int64? = nil
        var version: String? = nil
        var isHealthy = false
        var isSyncing = false
        var workingURL = url

        for scheme in ["https", "http"] {
            let tryURL = "\(scheme)://\(host):\(port)"
            let client = MasternodeClient(endpoint: tryURL, credentials: credentials)
            if let health = try? await withPeerTimeout(seconds: Self.healthTimeoutSecs) { try await client.healthCheck() } {
                blockHeight = health.blockHeight
                version = health.version
                isSyncing = health.isSyncing
                workingURL = tryURL

                if let expected = expectedGenesisHash,
                   let genesis = try? await withPeerTimeout(seconds: Self.healthTimeoutSecs, operation: { try await client.getGenesisHash() }) {
                    if genesis != expected {
                        await blacklistPeer(endpoint: tryURL)
                        blockHeight = nil
                        break
                    }
                }
                isHealthy = true
                break
            }
        }

        let wsAvailable = isHealthy ? await Self.checkWsAvailable(host: host, wsPort: port + 1) : false

        return PeerInfo(endpoint: workingURL, isActive: false, isHealthy: isHealthy,
                       wsAvailable: wsAvailable, pingMs: pingMs, blockHeight: blockHeight,
                       version: version, isSyncing: isSyncing)
    }

    private static func measureTcpPing(host: String, port: Int) async -> Int64? {
        await withCheckedContinuation { continuation in
            let start = Date()
            let conn = NWConnection(host: NWEndpoint.Host(host),
                                    port: NWEndpoint.Port(integerLiteral: UInt16(port)),
                                    using: .tcp)
            var done = false
            conn.stateUpdateHandler = { state in
                guard !done else { return }
                switch state {
                case .ready:
                    done = true
                    let ms = Int64(Date().timeIntervalSince(start) * 1000)
                    conn.cancel()
                    continuation.resume(returning: ms)
                case .failed, .cancelled:
                    done = true
                    continuation.resume(returning: nil)
                default: break
                }
            }
            conn.start(queue: .global())
            DispatchQueue.global().asyncAfter(deadline: .now() + 1) {
                guard !done else { return }
                done = true
                conn.cancel()
                continuation.resume(returning: nil)
            }
        }
    }

    private static func checkWsAvailable(host: String, wsPort: Int) async -> Bool {
        let ping = await measureTcpPing(host: host, port: wsPort)
        return ping != nil
    }

    private static func consensusFilter(peers: [PeerInfo], existingPeers: [PeerInfo] = []) -> [PeerInfo] {
        let allHeights = (peers + existingPeers).compactMap { $0.blockHeight }
        guard let bestHeight = allHeights.max() else { return peers }
        return peers.filter { p in
            guard let h = p.blockHeight else { return true }
            return (bestHeight - h) <= maxBlockLag
        }
    }

    private static func rankPeers(_ peers: [PeerInfo]) -> [PeerInfo] {
        peers.sorted {
            if $0.isHealthy != $1.isHealthy { return $0.isHealthy }
            if $0.isSyncing != $1.isSyncing { return !$0.isSyncing }
            if $0.wsAvailable != $1.wsAvailable { return $0.wsAvailable }
            let p0 = $0.pingMs ?? Int64.max
            let p1 = $1.pingMs ?? Int64.max
            return p0 < p1
        }
    }

    private static func fetchFromAPI(isTestnet: Bool, rpcPort: Int) async throws -> [String] {
        let urlString = isTestnet ? testnetAPIURL : mainnetAPIURL
        guard let url = URL(string: urlString) else { throw URLError(.badURL) }
        let (data, _) = try await URLSession.shared.data(from: url)
        guard let ips = try? JSONDecoder().decode([String].self, from: data) else {
            throw URLError(.cannotParseResponse)
        }
        guard !ips.isEmpty else { throw URLError(.zeroByteResource) }
        return ips.map { "https://\($0):\(rpcPort)" }
    }

    private static func saveCache(isTestnet: Bool, peers: [String], dir: URL) {
        let file = dir.appendingPathComponent(isTestnet ? "peers-testnet.json" : "peers-mainnet.json")
        try? JSONEncoder().encode(peers).write(to: file)
    }

    private static func loadCache(isTestnet: Bool, dir: URL) -> [String]? {
        let file = dir.appendingPathComponent(isTestnet ? "peers-testnet.json" : "peers-mainnet.json")
        guard let data = try? Data(contentsOf: file) else { return nil }
        return try? JSONDecoder().decode([String].self, from: data)
    }
}

private func withPeerTimeout<T>(seconds: Double, operation: @escaping () async throws -> T) async throws -> T {
    try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await operation() }
        group.addTask {
            try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            throw CancellationError()
        }
        guard let result = try await group.next() else { throw CancellationError() }
        group.cancelAll()
        return result
    }
}
